### PR TITLE
Upwind3rdOrderBiasedProductC2F 

### LINKED ISSUE #524: Third-order upwind FD scheme
Since a 1st-order upwind FD scheme is too dissipative (edit: typo), we want to add a 3rd-order upwind FD scheme. 

This can then be used with limiters in hybrid domains (both 2D hybrid domain, 1D CG x 1D FD, and 3D hybrid domain, 2D CG x 1D FD).

Here are a couple of references:
* Book: Hundsdorfer and Verwer - 2003 - Numerical Solution of Time-Dependent Advection-Diffusion-Reaction Equations (equations (3.27) & (3.28) for regular unstaggered upwind third-order finite differences stencil)
* Paper on staggered reconstruction: Pressel et al. - 2015 - Large-eddy simulation in an anelastic framework with closed
water and entropy balances [doi: 10.1002/2015MS000496](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2015MS000496) (equations (60)-(63))

### PR DESCRIPTION
**Pull request by valeriabarra**
This is a PR that adds high-order interpolation/reconstruction for the vertical flux on staggered 1D grids. 

A word of caution: Currently, we need to figure out the boundary conditions, so it supports only periodic meshes and, therefore, it's tested only on a periodic domain.

On periodic domains, for a constant velocity field, it does exhibit the desired order of convergence: 3. 
 
 This will close #524

### CODE REVIEW

File: test/Operators/finitedifference/column.jl [hunk 1]
```diff
@@ -6,243 +6,243 @@ import ClimaCore.Domains: Geometry
...
-        @test face_sum.a ≈ FT(pi)
-        @test face_sum.b ≈ FT(pi)
-    end
-end
+# @testset "PlusHalf" begin
+#     @test half + 0 == half
+#     @test half < half + 1
+#     @test half <= half + 1
+#     @test !(half > half + 1)
+#     @test !(half >= half + 1)
```
Comment (simonbyrne) [inline test/Operators/finitedifference/column.jl:14]: Uncomment lines

**Review by charleskawczynski — APPROVED** [review]
LGTM, just need to revert commenting the tests

**Comment by valeriabarra** [comment]
> LGTM, just need to revert commenting the tests

Done and rebased. Thanks!

**Pull request merged**